main_test.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"math/rand"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/google/uuid"
+)
+
+func TestGenerate(t *testing.T) {
+	uuid.SetRand(rand.New(rand.NewSource(1)))
+	in := configInput{
+		xdsServerUri:     "example.com:443",
+		gcpProjectNumber: 123456789012345,
+		vpcNetworkName:   "thedefault",
+		ip:               "10.9.8.7",
+		zone:             "uscentral-5",
+	}
+	expected := `{
+  "xds_servers": [
+    {
+      "server_uri": "example.com:443",
+      "channel_creds": [
+        {
+          "type": "google_default"
+        }
+      ]
+    }
+  ],
+  "node": {
+    "id": "52fdfc07-2182-454f-963f-5f0f9a621d72~10.9.8.7",
+    "cluster": "cluster",
+    "metadata": {
+      "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER": "123456789012345",
+      "TRAFFICDIRECTOR_NETWORK_NAME": "thedefault"
+    },
+    "locality": {
+      "zone": "uscentral-5"
+    }
+  }
+}`
+	out, err := generate(in)
+	if err != nil {
+		t.Fatalf("expected no error, got: %v", err)
+	}
+	if expected != string(out) {
+		t.Fatalf("\nexpected:\n%v\n----------------\ngot:\n%v",
+			expected, string(out))
+	}
+}
+
+func TestGetZone(t *testing.T) {
+	server := httptest.NewServer(nil)
+	defer server.Close()
+	overrideHttp(server)
```

Comment:
Optional: put the `NewServer` call in `overrideHttp` and have that return the server.

Reply:
I considered that and then had a reason not to do it. I think it was for clarity. I see the `nil` passed to NewServer is important, because it means we're using the default mux. That could be an argument to overrideHTTP() though. I think it was also that I wanted the server.close() to be more obvious; it isn't as obvious it needs to be shut down as a return value of overrideHTTP().

Reply:
It's `nil` in both places it is called, so the parameter doesn't actually seem all that important to the test as-is.  I'm fine with this either way, though.